wolfssl/internal.h

```diff
@@ -2599,8 +2601,79 @@ typedef enum {
 #ifdef WOLFSSL_QUIC
     TLSX_KEY_QUIC_TP_PARAMS_DRAFT   = 0xffa5, /* from draft-ietf-quic-tls-27 */
 #endif
+#if defined(HAVE_ECH)
+    TLSX_ECH                        = 0xfe0d,
+#endif
 } TLSX_Type;
 
+#if defined(HAVE_ECH)
+
+typedef enum
+{
+    ECH_TYPE_OUTER = 0,
+    ECH_TYPE_INNER = 1
+} EchType;
+
+typedef enum
+{
+    ECH_WRITE_GREASE,
+    ECH_WRITE_REAL,
+    ECH_WRITE_RETRY_CONFIGS,
+    ECH_WRITE_NONE,
+    ECH_PARSED_INTERNAL,
+} EchState;
+
+typedef struct EchCipherSuite
+{
+    word16 kdfId;
+    word16 aeadId;
+} EchCipherSuite;
+
+typedef struct EchConfig
+{
+    byte configId;
+    word16 kemId;
+    void* receiverPrivkey;
+    byte receiverPubkey[HPKE_Npk_MAX];
+    byte numCipherSuites;
+    EchCipherSuite* cipherSuites;
+    char* publicName;
+    byte* raw;
+    word32 rawLen;
+    struct EchConfig* next;
+} EchConfig;
+
+typedef struct ECH
+{
+    byte type;
+    word16 kemId;
+    EchCipherSuite cipherSuite;
+    byte configId;
+    byte enc[HPKE_Npk_MAX];
+    word16 encLen;
+    word32 innerClientHelloLen;
+    word32 paddingLen;
+    byte* innerClientHello;
+    byte* outerClientPayload;
+    const byte* aad;
+    word32 aadLen;
+    Hpke* hpke;
+    void* ephemeralKey;
+    EchState state;
+    EchConfig* echConfig;
+} ECH;
+
+WOLFSSL_LOCAL int EchConfigGetSupportedCipherSuite(EchConfig* config);
+
+WOLFSSL_LOCAL int TLSX_FinalizeEch(ECH* ech, byte* aad, word32 aadLen);
+
+WOLFSSL_API int wolfSSLGetEchConfig(EchConfig* config, byte* output,
+    word32* outputLen);
+
+WOLFSSL_API int wolfSSLGetEchConfigsEx(EchConfig* configs, byte* output,
```

Comment:
All new public facing API's (WOLFSSL_API) need additional negative tests, this would be a test like NULL input for each of the args in tests/api.c.

Reply:
Also our normal naming convention for public API's at the wolfSSL (TLS) layer is something like `wolfSSL_GetEchConfigs`. Do we need a `ex` version already? Can we accommodate just a single new API?

Reply:
these aren't supposed to be called from the public api, I thought everything in `internal.h` was for our code only to access. the ex version is necessary because the non ex version, which is public and in ssl.h, takes the ssl object from the user but when the `ECH_WRITE` function is called from `TLSX_Write` it doesn't have access to the ssl object so it passes the EchConfig directly. do I need to change the `WOLFSSL_API` term to something else and should I make the negative tests for the public functions below?
```WOLFSSL_API int wolfSSLGenerateEchConfig(WOLFSSL_CTX* ctx,
    const char* publicName);

WOLFSSL_API int wolfSSLSetEchConfigs64(WOLFSSL* ssl, char* echConfigs64,
    word32 echConfigs64Len);

WOLFSSL_API int wolfSSLSetEchConfigs(WOLFSSL* ssl, const byte* echConfigs,
    word32 echConfigsLen);

WOLFSSL_API int wolfSSLGetEchConfigs(WOLFSSL* ssl, byte* echConfigs,
    word32* echConfigsLen);
```

Reply:
Use `WOLFSSL_LOCAL`. It will prevent export of the function symbol.